internal/http/services/owncloud/ocs/response/response.go

```diff
@@ -34,13 +35,68 @@ type Response struct {
 
 // Payload combines response metadata and data
 type Payload struct {
-	XMLName struct{}      `json:"-" xml:"ocs"`
-	Meta    *ResponseMeta `json:"meta" xml:"meta"`
-	Data    interface{}   `json:"data,omitempty" xml:"data,omitempty"`
+	XMLName struct{}    `json:"-" xml:"ocs"`
+	Meta    *Meta       `json:"meta" xml:"meta"`
+	Data    interface{} `json:"data,omitempty" xml:"data,omitempty"`
 }
 
-// ResponseMeta holds response metadata
-type ResponseMeta struct {
+var (
+	metaStartElement = xml.StartElement{Name: xml.Name{Local: "meta"}}
+	dataName         = xml.Name{Local: "data"}
+	elementName      = xml.Name{Local: "element"}
+)
+
+// MarshalXML handles ocs specific wrapping of array members in 'element' tags for the data
+func (p Payload) MarshalXML(e *xml.Encoder, start xml.StartElement) (err error) {
+	// first the easy part
+	// use ocs as the surrounding tag
+	start.Name = xml.Name{Local: "ocs"}
+	if err = e.EncodeToken(start); err != nil {
+		return
+	}
+
+	// encode the meta tag
+	if err = e.EncodeElement(p.Meta, metaStartElement); err != nil {
+		return
+	}
+
+	// this is how to wrap theo data elements in their own <element> tag
+	render := func(v reflect.Value) {
+		if err = e.EncodeToken(dataName); err != nil {
+			return
+		}
+		for i := 0; i < v.Len(); i++ {
+			if err = e.EncodeElement(v.Index(i).Interface(), xml.StartElement{Name: elementName}); err != nil {
+				return
+			}
+		}
+		if err = e.EncodeToken(xml.EndElement{Name: dataName}); err != nil {
+			return
+		}
+	}
+
+	// we need to use reflection to determine if p.Data is an array or a slice
+	rt := reflect.TypeOf(p.Data)
```

Comment:
nothing against this, but I imagine this getting hit quite often, is there any way to avoid reflection here?

Reply:
type assertion perhaps?

```golang
package main

import (
	"fmt"
)

func main() {
	c := []string{"a", "b"}
	var a interface{} = c
	s, ok := a.([]string)
	fmt.Println(s, ok) // [a b] true
}
```

I don't have the entire problem scope on my head and I'm sure I have a blind spot as per why you came with this solution.

I would also add a benchmark perhaps, so we see how it performs. This can be done a posteriori perhaps?

Reply:
Reflection is evil, should be avoided, I'll create a issue for that.